File: project-1-calculator/calculator.cpp
// calculator.cpp

#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <string>

using namespace std;

int how_many_leading_digits(string s) {
    int i = 0;
    for (; i < s.length(); i++) {
        char c = s.at(i);
        if (!isdigit(c)) {
            return i;
        }
    }
    return i;
}

string leading_digits(string s) {
    int i = how_many_leading_digits(s);
    return s.substr(0, i);
}

vector<string> lex(string s) {
    int i = 0;
    vector<string> ts = vector<string>();
    while (i < s.length()) {
        char c = s.at(i);
        if (isdigit(c)) {
            string digits = leading_digits(s.substr(i));
            ts.push_back(digits);
            i += digits.length();
        } else {
            i++;
            if (c == ' ') {
                continue;
            } else if (c == '(' || c == ')' || c == '+' || c == '*' || c == '-' || c == '/') {
                ts.push_back(string(1, c));
            } else {
                cout << "Error: unexpected character, " << c << endl;
                exit(-1);
            }
        }
    }
    return ts;
}

bool is_num(string token) {
    for (int i = 0; i < token.length(); i++) {
        if (!isdigit(token.at(i))) return false;
    }
    return true;
}

bool is_operator(string token) {
    return token == "+" || token == "-"
        || token == "*" || token == "/";
}

bool is_lparen(string t) {
    return t == "(";
}

bool is_rparen(string t) {
    return t == ")";
}

int precedence(string op) {
    if (op == "+" || op  == "-") {
        return 1;
    } else if (op == "*" || op == "/") {
        return 2;
    } else {
        cout << "Impossible" << endl;
        exit(-1);
    }
}

queue<string> parse(vector<string> tokens) {
    // TODO: Implement this using the shunting yard algorithm.
    stack<string> ops = stack<string>();
    queue<string> output = queue<string>();
    for (int i = 0; i < tokens.size(); i++) {
        string t = tokens.at(i);
        if (is_num(t)) {
            output.push(t);
        } else if (is_operator(t)) {
            while (!ops.empty() && !is_lparen(ops.top()) && precedence(ops.top()) >= precedence(t)) {
                string e = ops.top();
                ops.pop();
                output.push(e);
            }
            ops.push(t);
        } else if (is_lparen(t)) {
            ops.push(t);
        } else if (is_rparen(t)) {
            while (!is_lparen(ops.top())) {
                string e = ops.top();
                ops.pop();
                output.push(e);
            }
        }
    }
    while (!ops.empty()) {
        string e = ops.top();
        ops.pop();
        output.push(e);
    }
    return output;
}

double apply(string op, double a, double b) {
    if (op == "+") return a+b;
    else if (op == "-") return a-b;
    else if (op == "*") return a*b;
    else if (op == "/") return a/b;
    cout << "Impossible: unknown operator" << endl;
    exit(-1);
}

double evaluate(queue<string> ast) {
    stack<double> buffer = stack<double>();
    while (!ast.empty()) {
        string t = ast.front();
        ast.pop();
        if (is_num(t)) {
            buffer.push(stod(t, nullptr));
        } else if (is_operator(t)) {
            double b = buffer.top();
            buffer.pop();
            double a = buffer.top();
            buffer.pop();
            double c = apply(t, a, b);
            buffer.push(c);
        }
    }
    return buffer.top();
}

int main() {
    string user_input;
    cout << "Input: ";
    getline(cin, user_input);
    vector<string> ts = lex(user_input);
    queue<string> ast = parse(ts);
    double result = evaluate(ast);
    cout << result << endl;
    return 0;
}

File: project-1-calculator/calculator.py
# calculator.py

def how_many_leading_digits(s):
    # Description: returns the number of leading digits in a string
    # Examples:
    #   1. how_many_leading_digits("") == 0
    #   2. how_many_leading_digits("hello") == 0
    #   3. how_many_leading_digits("2022") == 4
    #   4. how_many_leading_digits("20222022hello") == 8
    i = 0
    for c in s:
        if not c.isnumeric():
            return i
        i += 1
    return i

def leading_digits(s):
    # Description: returns the leading digits
    # Examples:
    #   1. leading_digits("") == ""
    #   2. leading_digits("hello") == ""
    #   3. leading_digits("2022") == "2022"
    #   4. leading_digits("20222022hello") == "20222022"
    i = how_many_leading_digits(s)
    return s[:i]

def lex(s):
    # Description: splits the string into "tokens"
    # Examples:
    #   1. lex("2022") == [2022]
    #   2. lex("1+1") == [1, '+', 1]
    #   3. lex("1*(2+3)") == [1, '*', '(', 2, '+', 3, ')']
    #   4. lex("1+2*(3+4)") == [1, '+', 2, '*', '(', 3, '+', 4, ')']
    i = 0
    ts = []
    while i < len(s):
        c = s[i]
        if c.isnumeric():
            digits = leading_digits(s[i:])
            ts.append(int(digits))
            i += len(digits)
        else:
            i += 1
            if c == ' ':
                continue # do nothing and skip to the next iteration
            elif c in ['(', ')', '+', '*', '-', '/']:
                # ^^^ or similarly, c == '(' or c == ')' or ... or c == '/'
                ts.append(c)
            else: # invalid character
                print('Error: unexpected character,', c)
                exit(-1) # Python function to exit the program early
    return ts

def is_number(t):
    return isinstance(t, int) 

def is_operator(t):
    return t in ['+','-','*','/']

def push(s, x):
    s.append(x)

def pop(s):
    x = s[-1]
    del s[-1]
    return x

def top(s):
    return s[-1]

def is_lparen(t):
    return t == '('

def empty(s):
    return len(s) == 0

def is_rparen(t):
    return t == ')'

# Push, pop, top

def precedence(t):
    if t in ['+', '-']: return 1
    elif t in ['*', '/']: return 2

def parse(ts):
    ops = []
    output = []
    for t in ts:
        if is_number(t):
            push(output, t)
        elif is_operator(t):
            while not empty(ops) and \
                    not is_lparen(t) and \
                     precedence(top(ops)) >= precedence(t):
                x = pop(ops)
                push(output, x)
            push(ops, t)
        elif is_lparen(t):
            push(ops, '(')
        elif is_rparen(t):
            while not is_lparen(top(ops)):
                x = pop(ops)
                push(output, x)
            pop(ops)
    while not empty(ops):
        x = pop(ops)
        push(output, x)
    return output

def run(t, a, b):
    if t == '+':  return a+b
    elif t == '-': return a-b
    elif t == '*': return a*b
    elif t == '/': return a/b

def evaluate(ast):
    s = []
    for t in ast:
        if isinstance(t, int):
            s.append(t)
        else: # + - * /
            a, b = s[-2], s[-1]
            del s[-2], s[-1]
            s.append(run(t, a, b))
    return s[0]

# The main program:
user_input = input('Input: ')
tokens = lex(user_input)
ast = parse(tokens)
result = evaluate(ast)
print(result)

